File: EAC/HUFF/example/LibExample.java
package EAC.HUFF.example;

import java.lang.foreign.Arena;
import java.lang.foreign.FunctionDescriptor;
import java.lang.foreign.Linker;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SymbolLookup;
import java.lang.foreign.ValueLayout;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Example Java program demonstrating how to use the EA Compression native
 * library
 * 
 * Usage: java LibExample.java <input file> <output file>
 */
public class LibExample {

    public static final int EA_FORMAT_HUFF = 0;
    public static final int EA_FORMAT_JDLZ = 1;
    public static final int EA_FORMAT_REF = 2;
    public static final int EA_FORMAT_BTREE = 3;
    public static final int EA_FORMAT_COMP = 4;
    public static final int EA_FORMAT_UNKNOWN = -1;

    public static String printFormat(int eaFormat) {
        return switch (eaFormat) {
            case EA_FORMAT_HUFF -> "EA_FORMAT_HUFF";
            case EA_FORMAT_JDLZ -> "EA_FORMAT_JDLZ";
            case EA_FORMAT_REF -> "EA_FORMAT_REF";
            case EA_FORMAT_BTREE -> "EA_FORMAT_BTREE";
            case EA_FORMAT_COMP -> "EA_FORMAT_COMP";
            default -> "EA_FORMAT_UNKNOWN";
        };
    }

    public static void main(String[] args) throws Throwable {
        if (args.length != 2) {
            System.err.println("Usage: java LibExample <input file> <output file>");
            System.exit(1);
        }

        String inputFile = args[0];
        String outputFile = args[1];

        if (inputFile.isEmpty() || outputFile.isEmpty()) {
            System.err.println("Input and output file paths must not be empty.");
            System.exit(1);
        }
        var pathToInput = Path.of(inputFile);
        var pathToOutput = Path.of(outputFile);

        if (!Files.exists(pathToInput) || !Files.isReadable(pathToInput)) {
            System.err.println("Input file does not exist or is not readable: " + inputFile);
            System.exit(1);
        }

        try (Arena arena = Arena.ofConfined()) {
            Linker linker = Linker.nativeLinker();
            SymbolLookup libLookup = SymbolLookup.libraryLookup("./libea_compression.so.1.0.0", arena);

            var ea_detect_format = linker.downcallHandle(
                    libLookup.find("ea_detect_format").orElseThrow(),
                    FunctionDescriptor.of(
                            ValueLayout.JAVA_INT,
                            ValueLayout.ADDRESS,
                            ValueLayout.JAVA_INT));

            var ea_get_decompressed_size = linker.downcallHandle(
                    libLookup.find("ea_get_decompressed_size").orElseThrow(),
                    FunctionDescriptor.of(
                            ValueLayout.JAVA_INT,
                            ValueLayout.ADDRESS,
                            ValueLayout.JAVA_INT));

            var ea_decompress = linker.downcallHandle(
                    libLookup.find("ea_decompress").orElseThrow(),
                    FunctionDescriptor.of(
                            ValueLayout.JAVA_INT,
                            ValueLayout.ADDRESS,
                            ValueLayout.JAVA_INT,
                            ValueLayout.ADDRESS,
                            ValueLayout.JAVA_INT));

            byte[] inputData = Files.readAllBytes(pathToInput);
            int dataSize = inputData.length;

            MemorySegment dataSegment = arena.allocateFrom(ValueLayout.JAVA_BYTE, inputData);

            int format = (int) ea_detect_format.invokeExact(
                    dataSegment,
                    dataSize);

            System.out.println("Detected format: " + printFormat(format));
            if (format == EA_FORMAT_UNKNOWN) {
                System.err.println("Unknown or unsupported EA compression format.");
                return;
            }

            int decompressedSize = (int) ea_get_decompressed_size.invokeExact(
                    dataSegment,
                    dataSize);

            System.out.println("Decompressed size: " + decompressedSize);

            MemorySegment outputSegment = arena.allocate(decompressedSize);

            int _ = (int) ea_decompress.invokeExact(
                    dataSegment,
                    dataSize,
                    outputSegment,
                    decompressedSize);

            byte[] outputData = new byte[decompressedSize];
            outputSegment.asByteBuffer().get(outputData);

            Files.write(pathToOutput, outputData);
            System.out.println("Decompression successful. Output written to: " + outputFile);
        }

    }
}

File: EAC/HUFF/example/lib-example.c
// Example usage of EA Compression Library
// Compile: gcc -o example example.c -lea_compression
// Or: gcc -o example example.c -L. -lea_compression

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "ea_compression_lib.h"

void print_format(ea_format_t format) {
    switch (format) {
        case EA_FORMAT_HUFF:   printf("HUFF\n"); break;
        case EA_FORMAT_JDLZ:   printf("JDLZ\n"); break;
        case EA_FORMAT_REF:    printf("REF\n"); break;
        case EA_FORMAT_BTREE:  printf("BTREE\n"); break;
        case EA_FORMAT_COMP:   printf("COMP\n"); break;
        default:               printf("UNKNOWN\n"); break;
    }
}

int main(int argc, char *argv[]) {
    printf("EA Compression Library Example\n");
    printf("Version: %s\n\n", ea_version());

    if (argc != 3) {
        printf("Usage: %s <compress|decompress> <file>\n", argv[0]);
        return 1;
    }

    const char *mode = argv[1];
    const char *filename = argv[2];

    FILE *file = fopen(filename, "rb");
    if (!file) {
        printf("Error: Cannot open file '%s'\n", filename);
        return 1;
    }

    // Get file size
    fseek(file, 0, SEEK_END);
    int file_size = ftell(file);
    rewind(file);

    // Read file
    unsigned char *file_data = malloc(file_size);
    if (!file_data) {
        printf("Error: Cannot allocate memory\n");
        fclose(file);
        return 1;
    }
    fread(file_data, 1, file_size, file);
    fclose(file);

    if (strcmp(mode, "decompress") == 0) {
        // Decompress mode
        printf("Detecting format...\n");
        ea_format_t format = ea_detect_format(file_data, file_size);
        printf("Format: ");
        print_format(format);

        if (format == EA_FORMAT_UNKNOWN) {
            printf("Error: Unknown compression format\n");
            free(file_data);
            return 1;
        }

        int decompressed_size = ea_get_decompressed_size(file_data, file_size);
        if (decompressed_size <= 0) {
            printf("Error: Cannot determine decompressed size\n");
            free(file_data);
            return 1;
        }

        printf("Compressed size: %d bytes\n", file_size);
        printf("Decompressed size: %d bytes\n", decompressed_size);

        unsigned char *decompressed = malloc(decompressed_size);
        if (!decompressed) {
            printf("Error: Cannot allocate memory for decompression\n");
            free(file_data);
            return 1;
        }

        printf("Decompressing...\n");
        int result = ea_decompress(file_data, file_size, decompressed, decompressed_size);
        
        if (result > 0) {
            printf("Success! Decompressed %d bytes\n", result);
            
            // Write output
            char outname[256];
            snprintf(outname, sizeof(outname), "%s.decompressed", filename);
            FILE *outfile = fopen(outname, "wb");
            if (outfile) {
                fwrite(decompressed, 1, result, outfile);
                fclose(outfile);
                printf("Output written to: %s\n", outname);
            }
        } else {
            printf("Error: Decompression failed with code %d\n", result);
        }

        free(decompressed);

    } else if (strcmp(mode, "compress") == 0) {
        // Compress mode (example with HUFF)
        printf("Compressing with HUFF format...\n");
        printf("Input size: %d bytes\n", file_size);

        int max_compressed = file_size * 2 + 16;
        unsigned char *compressed = malloc(max_compressed);
        if (!compressed) {
            printf("Error: Cannot allocate memory for compression\n");
            free(file_data);
            return 1;
        }

        int result = ea_compress_huff(file_data, file_size, compressed, max_compressed, 0);
        
        if (result > 0) {
            printf("Success! Compressed to %d bytes (%.2f%% of original)\n", 
                   result, (result * 100.0) / file_size);
            
            // Write output
            char outname[256];
            snprintf(outname, sizeof(outname), "%s.huff", filename);
            FILE *outfile = fopen(outname, "wb");
            if (outfile) {
                fwrite(compressed, 1, result, outfile);
                fclose(outfile);
                printf("Output written to: %s\n", outname);
            }
        } else {
            printf("Error: Compression failed with code %d\n", result);
        }

        free(compressed);

    } else {
        printf("Error: Invalid mode. Use 'compress' or 'decompress'\n");
        free(file_data);
        return 1;
    }

    free(file_data);
    return 0;
}
